### PR TITLE
Nifi-1972 Apache Ignite Put Cache Processor

### PR DESCRIPTION
**Pull request by mans2singh**
Nifi Put Ignite cache processor which streams data into Ignite Cache.

### CODE REVIEW

**Comment by mans2singh** [comment]
Hey Folks:

Please let me know your thoughts/suggestions on this Nifi Ignite Put Processor.

Thanks

**Comment by pvillard31** [comment]
@mans2singh Could you check all the lines marked as removed? I believe there should not be any and this is probably because of some settings in your IDE.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+/**
+ * Base class for Ignite processors
+ */
+public abstract class AbstractIgniteProcessor extends AbstractProcessor  {
+
+    /**
+     * Ignite spring configuration file
+     */
+    public static final PropertyDescriptor IGNITE_CONFIGURATION_FILE = new PropertyDescriptor.Builder()
+        .name("Ignite Spring Properties Xml File")
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java]: Could you use both .name() and displayName() for backward compatibility?

See https://mail-archives.apache.org/mod_mbox/nifi-dev/201605.mbox/%3C5A6FDF1E-1889-46FE-A3C4-5D2F0A905979@apache.org%3E

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+    })
+public class PutIgniteCache extends AbstractIgniteCacheProcessor {
+
+    /**
+     * The batch size of flow files to be processed on invocation of onTrigger
+     */
+    public static final PropertyDescriptor BATCH_SIZE = new PropertyDescriptor.Builder()
+            .name("Batch Size For Entries")
+            .description("Batch size for entries (1-500).")
+            .defaultValue("250")
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: What are the reasons for this default value and the range 1-500?

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+            .addValidator(StandardValidators.createLongValidator(1, 10, true))
+            .sensitive(false)
+            .build();
+
+    /**
+     * Data streamers per node buffer size
+     */
+    public static final PropertyDescriptor DATA_STREAMER_PER_NODE_BUFFER_SIZE = new PropertyDescriptor.Builder()
+            .name("Data Streamer Per Node Buffer Size")
+            .description("Data streamer per node buffer size (1-500).")
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: Buffer size? Shouldn't we use a data unit? Or does it mean something else?

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+            .build();
+
+    /**
+     * Data streamers override values property
+     */
+    public static final PropertyDescriptor DATA_STREAMER_ALLOW_OVERRIDE = new PropertyDescriptor.Builder()
+            .name("Data Streamer Allow Override")
+            .description("Whether to override values already in the cache")
+            .defaultValue("false")
+            .allowableValues(new AllowableValue("true"), new AllowableValue("false"))
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: I believe that when there is a default and allowable values, we want to set .required(true)

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final int batchSize = context.getProperty(BATCH_SIZE).asInteger();
+        final List<FlowFile> flowFiles = session.get(batchSize);
+
+        if (flowFiles.isEmpty()) {
+            return;
+        }
+
+//        Map<String, byte[]> cacheMap = new java.util.HashMap<>();
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: to remove?

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/AbstractIgniteCacheProcessor.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * Base class of Ignite cache based processor
+ */
+public abstract class AbstractIgniteCacheProcessor extends AbstractIgniteProcessor {
+
+    /**
+     * Flow File attribute for cache entry key
+     */
+    public static final String IGNITE_CACHE_ENTRY_KEY = "ignite.cache.entry.key";
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/AbstractIgniteCacheProcessor.java]: Wouldn't be better to let the user decide what attribute is used as the key in Ignite? I believe this is what we do in PutDistributedMapCache processor and it would avoid an unnecessary step with UpdateAttribute processor.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+                            StreamUtils.fillBuffer(in, byteArray, true);
+                        }
+                    });
+
+                    cacheItems.add(new AbstractMap.SimpleEntry<String,byte[]>(key, byteArray));
+                    successfulFlowFiles.add(flowFile);
+
+                } catch (Exception e) {
+                    getLogger().error("Failed to insert {} into IgniteDB due to {}", new Object[] { flowFile, e }, e);
+                    session.transfer(flowFile, REL_FAILURE);
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: Would it be interesting to yield the processor here?

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+        } finally {
+            if (!cacheItems.isEmpty()) {
+                IgniteFuture<?> futures = igniteDataStreamer.addData(cacheItems);
+                Object result = futures.get();
+                getLogger().debug("Result {} of addData", new Object [] {result});
+            }
+
+            if (!successfulFlowFiles.isEmpty()) {
+                successfulFlowFiles = updateSuccessfulFlowFileAttributes(flowFiles, successfulFlowFiles, session);
+                session.transfer(successfulFlowFiles, REL_SUCCESS);
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: What about provenance events?

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        .description("Ignite spring configuration file, <path>/<ignite-configuration>.xml")
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java]: Shouldn't be required to be sure there is no surprise for the user? At least, I think you could add that if this value is not set, a default configuration is used in order to bind with a local Ignite endpoint on 127.0.0.1:47500..47509?

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/resources/ignite-client.xml [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+  limitations under the License.
+-->
+<beans xmlns="http://www.springframework.org/schema/beans"
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:schemaLocation="http://www.springframework.org/schema/beans
+        http://www.springframework.org/schema/beans/spring-beans.xsd">
+    <!-- Imports default Ignite configuration -->
+    <import resource="ignite-default-client.xml"/>
+
+    <bean parent="ignite.cfg"/>
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/resources/ignite-client.xml]: What is ignite.cfg for?

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java (hunk 1 above)
Comment (mans2singh) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java]: I will do that @pvillard31

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java (hunk 2 above)
Comment (mans2singh) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: I just picked a number for processing in a single batch to avoid buffering too many objects in memory.  If you have any advice based on your experience, I can change it.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java (hunk 3 above)
Comment (mans2singh) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: The buffer size is the number of objects to buffer.  Let me know what you think.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java (hunk 4 above)
Comment (mans2singh) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: I've added required(true) as you had recommended, I think is a safer choice.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java (hunk 5 above)
Comment (mans2singh) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: You are right, have removed.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/AbstractIgniteCacheProcessor.java (hunk 6 above)
Comment (mans2singh) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/AbstractIgniteCacheProcessor.java]: Will update it.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java (hunk 7 above)
Comment (mans2singh) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: Will add yield here.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java (hunk 8 above)
Comment (mans2singh) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: The flow files are being processed in batches and I could not find a send method on provenance reporter that takes a batch of flow files. If there is a better way of handling this, please let me know.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java (hunk 9 above)
Comment (mans2singh) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java]: Ok will add that to the documentation.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/resources/ignite-client.xml (hunk 10 above)
Comment (mans2singh) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/resources/ignite-client.xml]: This is the id of the abstract IgniteConfiguration bean (defined in the ignite-default-client.xml) so that the user can override it's properties.

**Comment by mans2singh** [comment]
Hi @pvillard31 - I've made the changes that you've recommended except for the provenance since I could not find any method to support a collection of flow files.  Please let me know if I have missed anything or you have additional suggestions for me.  Thanks

**Comment by pvillard31** [comment]
Thanks @mans2singh ! I'll have a look asap (probably on Sunday or next week). Regarding the provenance event, I think we could have a SEND event for each flow file successfully sent to Ignite. Thoughts?

Note: it seems that your PR has conflicts against master.

**Comment by mans2singh** [comment]
@pvillard31 - I messed up the merge may have to recreate the branch.  Let me fix that and then you can review it. 
Thanks

**Comment by mans2singh** [comment]
@pvillard31 - I've added provenance and corrected the merge and conflict issue. Please let me know your thoughts.

**Comment by pvillard31** [comment]
@mans2singh 
Just had a new look and ran a workflow to test the processor. Few remarks even though I am not familiar with Ignite... First, I am not sure to see how I can check the content of what is sent to Ignite, do you have recommendation about that?

Then, it seems that if I don't provide a configuration file, the processor is running an embedded Ignite cache but in that case I am not sure of what is the interest for the user. Even with a local Ignite node running it does not seem to connect to it, normal?

Other remark, in the logs, I see some periodic status about the Ignite instance
````
2016-06-14 22:16:25,694 INFO [grid-timeout-worker-#23%null%] org.apache.ignite.internal.IgniteKernal 
Metrics for local node (to disable set 'metricsLogFrequency' to 0)
    ^-- Node [id=6b53a263, name=null, uptime=00:01:00:002]
    ^-- H/N/C [hosts=2, nodes=2, CPUs=16]
    ^-- CPU [cur=0.67%, avg=0.79%, GC=0%]
    ^-- Heap [used=260MB, free=49.12%, comm=512MB]
    ^-- Non heap [used=119MB, free=-1%, comm=125MB]
    ^-- Public thread pool [active=0, idle=7, qSize=0]
    ^-- System thread pool [active=0, idle=16, qSize=0]
    ^-- Outbound messages queue [size=0]
````
I don't know if it makes sens to have this kind of logs in NiFi logs, thoughts?

Regarding the transit URI of the provenance event, I am not convinced. Is there a way to get IP/port of the node the processor is connected to and use this information in the event? (like, I believe, it is done in Kafka processors)

I believe there is also some kind of issues when stopping the processor: it doesn't seem to stop gracefully and I get some stack traces when trying to run it again...  in the end, none of the incoming flow files are ingested. Here is the stack trace if it helps (https://gist.github.com/pvillard31/d2e97008b36b4cd9edcfec8d5ed20b86#file-putignite-log)

**Comment by apiri** [comment]
Hey @mans2singh,

A bit late to the game but was doing some scanning over issues and performing cleanup.

Just curious if any thought was given to treating this as an alternate implementation for a DistributedCache controller service.  Such an implementation would allow existing processors to make use of this backend as well.  Also not overly familiar with Ignite, so if there are any gotchas/caveats that would prevent this apologies for the lack of knowledge on my behalf.

**Comment by mans2singh** [comment]
Hey @pvillard31 - Thanks for your comments and I will try to address to your questions below:

1.  In order to run integration tests, we need to run apache ignite 1.6 locally (or create the proper xml configuration file - just like under the src/test/resources directory).  
2. There are two integration tests (and a suite of unit tests) with the module.  The integration tests validate the data the contents of the flow with the data in the ignite cache.
3. If the config file is not provided the processor creates a ignite client which tries to discover the ignite cluster on the localhost.  This could be used in case a single Nifi node is collocated with Ignite Cache node. When I run the processor using Nifi application, I can see the entry count corresponding the flow files in ignite cache by using the ignitevisorcmd.sh application.  If you see any problems with it, please let me know and I will to reproduce the scenario locally.
4. Regarding provenance, Ignite tries to shield the clients from the servers by using a discovery spi.  While we can find out all the remote or local nodes in the cluster using this spi, we do not know which nodes will host the entry.  Getting all the nodes in the cluster will also impact the performance if we do it for the provenance of each flow file.  So, I've changed provenance to provide cache name which is how the client should access the entries.  Let me know your thoughts on it.
5. The logs for Ignite can be handled by changing the log levels and I will look into it.
6. I've been able to start/stop the cache and don't see the exceptions you've mentioned. If you have specific steps to reproduce them, please let me know.

@apiri - My thought was that this processor is a egress component for streaming large volume/velocity data to the ignite cache.  If you have any thought/recommendation for me, please let me know.

Thanks for all your advice.

**Comment by mans2singh** [comment]
Hi @pvillard31 @apiri - Just wanted to see if you have any thoughts/recommendations for me.  
Also, @pvillard31 were you able to run the integration tests and validate that the entries were saved in cache ?
Mans

**Comment by pvillard31** [comment]
Hi @mans2singh, I didn't find time to give it a new try, but I hope I will this week.

**Comment by mans2singh** [comment]
Hey @pvillard31 @aldrin - Just wanted to check if you have any suggestions/recommendations for this pull request.
Thanks

**Comment by pvillard31** [comment]
Hey @mans2singh, I'm really busy this week and I won't be able to have a look before next week :(
In the meantime, you can fix the conflicts against master. And maybe @apiri will have a chance to have a look before I do.

**Comment by mans2singh** [comment]
Hello @pvillard31 @apiri - I've resolved the conflicts for the pull request.  Please let me know if you have any additional feedback/recommendations for me.  Thanks

File: nifi-assembly/pom.xml [hunk 11]
```diff
@@ -379,7 +384,7 @@ language governing permissions and limitations under the License. -->
         <nifi.state.management.embedded.zookeeper.properties>./conf/zookeeper.properties</nifi.state.management.embedded.zookeeper.properties>
         <nifi.state.management.provider.local>local-provider</nifi.state.management.provider.local>
         <nifi.state.management.provider.cluster>zk-provider</nifi.state.management.provider.cluster>
-        
+
```
Comment (pvillard31) [inline nifi-assembly/pom.xml]: Could you revert unnecessary change here?

File: nifi-assembly/pom.xml [hunk 12]
```diff
@@ -425,7 +430,7 @@ language governing permissions and limitations under the License. -->
         <nifi.provenance.repository.query.threads>2</nifi.provenance.repository.query.threads>
         <nifi.provenance.repository.index.threads>1</nifi.provenance.repository.index.threads>
         <nifi.provenance.repository.compress.on.rollover>true</nifi.provenance.repository.compress.on.rollover>
-        <nifi.provenance.repository.indexed.fields>EventType, FlowFileUUID, Filename, ProcessorID, Relationship</nifi.provenance.repository.indexed.fields> 
+        <nifi.provenance.repository.indexed.fields>EventType, FlowFileUUID, Filename, ProcessorID, Relationship</nifi.provenance.repository.indexed.fields>
```
Comment (pvillard31) [inline nifi-assembly/pom.xml]: Could you revert unnecessary change here?

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/pom.xml [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+       </dependency>
+       <!--
+       <dependency>
+           <groupId>org.apache.ignite</groupId>
+           <artifactId>ignite-core</artifactId>
+           <type>test-jar</type>
+           <scope>test</scope>
+       </dependency>
+     -->
+       <dependency>
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/pom.xml]: If not needed, could we remove it?

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+public abstract class AbstractIgniteProcessor extends AbstractProcessor  {
+
+    /**
+     * Ignite spring configuration file
+     */
+    public static final PropertyDescriptor IGNITE_CONFIGURATION_FILE = new PropertyDescriptor.Builder()
+        .displayName("Ignite Spring Properties Xml File")
+        .name("ignite-spring-properties-xml-file")
+        .description("Ignite spring configuration file, <path>/<ignite-configuration>.xml. If the " +
+            "configuration file is not provided, default Ignite configuration is used which default " +
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java]: Wording? "is used" twice?

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    /**
+     * Initialize ignite instance
+     * @param context process context
+     */
+    public void initializeIgnite(ProcessContext context) {
+
+        if ( getIgnite() != null ) {
+            getLogger().info("Ignite already initilaized");
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java]: Typo. Also, do we want this with info level instead of debug? just asking in case we could have this message a lot...

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/AbstractIgniteCacheProcessor.java [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+    /**
+     * Ignite cache name
+     */
+    protected static final PropertyDescriptor CACHE_NAME = new PropertyDescriptor.Builder()
+            .displayName("Ignite Cache Name")
+            .name("ignite-cache-name")
+            .description("The name of the ignite cache where e")
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/AbstractIgniteCacheProcessor.java]: "where e"?

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/AbstractIgniteCacheProcessor.java (hunk 6 above)
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/AbstractIgniteCacheProcessor.java]: Is it still needed? I believe you created a property with the same name in the processor.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java [hunk 17]
```diff
@@ -0,0 +1,393 @@
...
+     * Check if flow if corrupted (either flow file is empty or does not have a key attribute)
+     * @param flowFile the flow file to check
+     * @param key the cache key
+     * @return <code>true</code> if flow file is incomplete
+     */
+    private boolean isFailedFlowFile(FlowFile flowFile, String key) {
+        if ( StringUtils.isEmpty(key) ) {
+            return true;
+        }
+        if ( flowFile.getSize() == 0 ) {
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]:
````java
return flowFile.getSize() == 0;
````

File: nifi-nar-bundles/pom.xml [hunk 18]
```diff
@@ -55,18 +55,19 @@
         <module>nifi-scripting-bundle</module>
         <module>nifi-elasticsearch-bundle</module>
         <module>nifi-amqp-bundle</module>
-	    <module>nifi-splunk-bundle</module>
+	      <module>nifi-splunk-bundle</module>
```
Comment (pvillard31) [inline nifi-nar-bundles/pom.xml]: Could you revert unnecessary changes on this file?

File: pom.xml [hunk 19]
```diff
@@ -1,13 +1,13 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!-- Licensed to the Apache Software Foundation (ASF) under one or more contributor
```
Comment (pvillard31) [inline pom.xml]: Could you revert unnecessary changes on this file?

**Comment by pvillard31** [comment]
@mans2singh I added some minor comments. You also need to add a NOTICE file for licensing aspects in the nar project (you may want to have a look to other processors).

I just tried to run the processor on my laptop with a local running Ignite instance (1.6.0). It seems to behave as expected. I didn't run into issues like I did the last time. Regarding the comments/suggestions you made, I'm not very familiar with Ignite and I am not the best one to judge. I think that once all remarks are addressed we can get this merged into master and see if changes are needed depending on how this processor is used.

However, regarding point 4 (logs) did you make any progress? I am not sure we want all the Ignite client logs into the NiFi log file.

**Comment by mans2singh** [comment]
Hi @pvillard31 - 

Thanks for your thorough feedback. I've updated the code based on your comments (removed commented dependency in pom file, added license and notice files under nar project, corrected typos in documentation and log message, changed log level and added log level warn in logback.xml file to avoid ignite info message).
  
Please let me know if you have any other comments/recommendations.

Thanks again for your help.

**Comment by mans2singh** [comment]
@pvillard31 - I am checking into the build failures.

**Comment by mans2singh** [comment]
@pvillard31 - The build is passing locally for me, so I am not sure what could the issue with the travis build.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-nar/src/main/resources/META-INF/NOTICE [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+    The following NOTICE information applies:
+      Apache Ignite
+      Copyright 2015 The Apache Software Foundation
+
+  (ASLv2) Apache Commons IO
+    The following NOTICE information applies:
+      Apache Commons IO
+      Copyright 2002-2012 The Apache Software Foundation
+
+  (ASLv2) Guava
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-nar/src/main/resources/META-INF/NOTICE]: Since this is only a test scope dependency and not bundled into the nar, this does not need to be a part of the notice

File: pom.xml [hunk 21]
```diff
@@ -227,6 +227,29 @@ language governing permissions and limitations under the License. -->
...
+                <groupId>org.apache.ignite</groupId>
+                <artifactId>ignite-spring</artifactId>
+                <version>1.6.0</version>
+            </dependency>
+            <dependency>
+                <groupId>org.apached.ignite</groupId>
+                <artifactId>ignite-core</artifactId>
+                <version>1.6.0</version>
+                <type>test-jar</type>
+                <scope>test</scope>
```
Comment (JPercivall) [inline pom.xml]: I'd suggest against providing the scope here and instead putting it in the pom where it's used. Adding the scope test here will propagate to any package that brings in this dependency (even transitively).

Comment also applies to ignite-log4j2

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-nar/src/main/resources/META-INF/NOTICE [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+This product includes software developed at
+The Apache Software Foundation (http://www.apache.org/).
+
+===========================================
+Apache Software License v2
+===========================================
+
+The following binary components are provided under the Apache Software License v2
+
+  (ASLv2) Apache Ignite
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-nar/src/main/resources/META-INF/NOTICE]: This notice should be this: https://github.com/apache/ignite/blob/1.6.0/NOTICE

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-nar/src/main/resources/META-INF/NOTICE [hunk 23]
```diff
@@ -0,0 +1,26 @@
+nifi-ignite-nar
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-nar/src/main/resources/META-INF/NOTICE]: Any NOTICE information in this Notice file that is not in the nifi-assembly NOTICE,  should be added to the assembly NOTICE as well

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/logback.xml [hunk 24]
```diff
@@ -89,7 +89,8 @@
     <logger name="org.apache.nifi.processors" level="WARN"/>
     <logger name="org.apache.nifi.processors.standard.LogAttribute" level="INFO"/>
     <logger name="org.apache.nifi.controller.repository.StandardProcessSession" level="WARN" />
-    
+
+    <logger name="org.apache.ignite" level="WARN"/>
```
Comment (JPercivall) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/logback.xml]: Instead of adding this to the default logback, which would get used for every NiFi instance, why not add documentation to the description or Additional Details section stating that periodic stats are logged and this statement, added to the logback.xml, will suppress them. 

This will get around adding a logger definition here that's specific to the processor but also give the user the ability to change it if needed. Besides, some users may like having the stats logged by default.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-nar/src/main/resources/META-INF/NOTICE (hunk 22 above)
Comment (mans2singh) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-nar/src/main/resources/META-INF/NOTICE]: Updated NOTICE.

**Comment by mans2singh** [comment]
Hi @JPercivall - I've updated the code based on your comments.  Please let me know if you have any other comments.  Thanks.

**Comment by JPercivall** [comment]
Hey @mans2singh, no other comments. Just those couple comments regarding the notice/license and the logging, and it looks like they were addressed. I'll defer to @pvillard31 to finish the review.

File: nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java [hunk 25]
```diff
@@ -0,0 +1,392 @@
...
+@EventDriven
+@SupportsBatching
+@Tags({ "Ignite", "insert", "update", "stream", "write", "put", "cache", "key" })
+@InputRequirement(Requirement.INPUT_REQUIRED)
+@CapabilityDescription("Stream the contents of a FlowFile to Ignite Cache using DataStreamer. " +
+    "The processor uses the value of FlowFile attribute (Ignite cache entry key) as the " +
+    "cache key and the byte array of the FlowFile as the value of the cache entry value.  Both the string key and a " +
+    " non-empty byte array value are required otherwise the FlowFile is transfered to the failure relation. " +
+    "Note - The Ignite Kernel periodically outputs node perforance statistics to the logs. This message " +
+    " can be turned off by setting the log level for logger 'org.apache.ignite' to WARN.")
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/PutIgniteCache.java]: typo: "performance"
also I'd say "... to WARN in logback.xml configuration file."

**Comment by pvillard31** [comment]
Thanks @JPercivall !
@mans2singh. It LGTM, one last comment about wording and also could you confirm my old comment in ``AbstractIgniteCacheProcessor`` about ``IGNITE_CACHE_ENTRY_KEY``?
Once we're good, I'll merge into master. Thanks!

**Comment by mans2singh** [comment]
Hi @pvillard31 - 

I've removed the ignite key constant from the abstract cache processor.  I missed your comment about it earlier.  I've also updated the documentation for performance log outputs as you had recommended.

Let me know if there is anything else outstanding.

Thanks for your and @JPercivall feedback.

Mans

**Comment by pvillard31** [comment]
@mans2singh Thanks for your work, it LGTM! Would you mind rebasing your commits?

**Comment by mans2singh** [comment]
@pvillard31 - I've rebased the commits.  Let me know if you have any other comments/recommendations.

Thanks

**Comment by pvillard31** [comment]
Sorry I meant to rebase and squash them into one single commit ;)

**Comment by mans2singh** [comment]
@pvillard31 - I've squashed the commits into a single one.  Thanks

**Comment by pvillard31** [comment]
@mans2singh I was about to merge but there are checkstyle issues:

````
[WARNING] src/test/java/org/apache/nifi/processors/ignite/cache/ITPutIgniteCache.java[41:1] (whitespace) FileTabCharacter: Line contains a tab character.
[WARNING] src/test/java/org/apache/nifi/processors/ignite/cache/ITPutIgniteCache.java[129] (regexp) RegexpSinglelineJava: Line has trailing whitespace.
````

Could you take care of this?

**Comment by mans2singh** [comment]
@pvillard31 - 

I've corrected checkstyle issues.  Please let me know if you see any other issue.  

Thanks again.

Mans
